File: 56-merge-intervals/56-merge-intervals.js
/**
 * @param {number[][]} intervals
 * @return {number[][]}
 */
var merge = function(intervals) {
    intervals = intervals.sort((a,b)=>a[0]-b[0]);
    intervals.push([Infinity,Infinity])
    
    let res = [];
    let prev;
    for(let interval of intervals){
        if(!prev){
            prev = interval;
            continue;
        }
        
        if(prev[1]<interval[0]){
            res.push(prev);
            prev = interval;
            continue;
        } else if(prev[1]<interval[1]){
            prev[1] = interval[1];
        }
    }
    
    return res;
};

File: 56-merge-intervals/56-merge-intervals.py
class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        intervals.sort(key=lambda a: a[0])
        stack = deque([intervals[0]])
        
        for i in range(1,len(intervals)):
            if stack[-1][1]>=intervals[i][0]:
                stack[-1] = [stack[-1][0],max(stack[-1][1],intervals[i][1])]
            else:
                stack.append(intervals[i])
        return [*stack]